cli/src/main.rs

```diff
@@ -6,25 +6,46 @@
 
 mod commands;
 use clap::Parser as ClapParser;
+use clap::{crate_authors, crate_version, ColorChoice, Parser};
+use subxt_codegen::utils::Uri;
 
 /// Subxt utilities for interacting with Substrate based nodes.
 #[derive(Debug, ClapParser)]
-enum Command {
-    Metadata(commands::metadata::Opts),
-    Codegen(commands::codegen::Opts),
-    Compatibility(commands::compatibility::Opts),
-    Version(commands::version::Opts),
+pub enum SubCommand {
+    Metadata(commands::metadata::MetadataOpts),
+    Codegen(commands::codegen::CodegenOpts),
+    Compatibility(commands::compatibility::CompatOpts),
+    Version(commands::version::Version),
+}
+
+#[derive(Debug, Parser)]
+#[clap(version = crate_version!(), author = crate_authors!(), color=ColorChoice::Always)]
+pub struct CliOpts {
+    #[clap(subcommand)]
+    pub subcmd: SubCommand,
+
+    /// The url of the substrate node connect to
+    #[clap(
+        name = "url",
+        long,
+        value_parser,
+        default_value = "http://localhost:9933",
+        env = "SUBXT_URL"
+    )]
+    url: Uri,
```

Comment:
I don't like this being moved to be a global arg; yes, perhaps at the moment eac hsubcommand requires a URL, but that definitely need not always be the case (eg why should a `version` commad need a url).

I'd much rather that each command specifies every option it cares about, and I don't care if that happens to lead to some code duplication because I think it also makes it easier to maintain :)

Reply:
(something that you _can_ do I think is to have a custom `Uri` type that has these clap attrs on and such so that you can just use the custom type in each command where we want it. I'd prefer that sort of thing (assuming it's actually possible as I described it)

Reply:
Sure not all the commands may need a url but it is unlikely that the core of `subxt` will work without a URL.
With the URL as part of each command, it is rather hard to inject the URL when it is not the default.

Making the URL global makes it available to all commands, they don't have to use it.

It makes the experience for users and devs also nicer as you can define the ENV once and then keep using commands without specifying the `--url` in various places.

Even without the ENV, it allows using a command that always start with `subxt --url <foobar> <the command>` and users can easily recall their command history and run another command by changing only the command part.

Reply:
Won't the ENV var part work if the variable is part of some subcommand? (and in that case, if somebody always wants to have some URL set then they can just set the ENV var and not worry about the `--url` at all).

And I don't feel that being able to put `--url` first is worth it anyway (also, iirc it also _requires_ that URL comes first, unlike all other params which must follow the subcommand, and I don't like that either; it's not consistent).

So I am not convinced I'm afraid!

Reply:
I am with you for things like `file` since most of the commands do not need a file (if that last assumption is wrong, then file should be IMO global as well). I am basing this PR on the fact that `subxt` is a tool that mostly will need to connect to an endpoint (besides a few utility commands such as version, help, etc..).

If the args are local to the commands, you cannot call the following: `subxt --url <foo> <command_1>`
and you must pass: `subxt <command_1> --url <foo>`. 

Having `--url` global however allows both and the support as ENV make the flag vanish from the user's view.

`cargo` for instance has lots of global level ENV such as `RUSTDOCFLAGS` that is totally irrelevant to some of the cargo commands such as `cargo --version` and if the user would happen to provide the flags or have the ENV set, this is simply ignored.